akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -74,11 +76,38 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
 
   private def updateState(streamId: Int, handle: IncomingStreamState => IncomingStreamState): Unit = {
     val oldState = streamFor(streamId)
-    val newState = handle(oldState)
+    val newState: IncomingStreamState = handle(oldState)
+
+    // only validate maxConcurrentStreams compliance if we'd increment the number
+    // of streams in one of the open states. This way, when the value of SETTINGS_MAX_CONCURRNET_STREAMS
+    // decreases, inflight streams can still finish and close.
+    val needsMaxConcurrentComplianceCheck = (oldState, newState) match {
+      case (Idle, _) => true
+      case _         => false
+    }
+
+    if (needsMaxConcurrentComplianceCheck) {
+      // incomingStreams contains all the open streams and an extra one that's Idle9
+      val currentConcurrentStreams = incomingStreams.size - 1
```

Comment:
Not sure why you would remove that `Idle` one? It should only Idle because you haven't applied the state update further down...

Reply:
This was an off-by-one error in concordance with the condition. Fixed the condition, fixed the subtraction.